### ISSUE TITLE
Question about Sequencer.lua 

### ISSUE THREAD

**Issue by helson73** [OpenNMT/OpenNMT#61]
I've implemented recursive net, and initialize sequencer with that. (also memory optimizer)
Source code is
```lua
require('nngraph')
local RVNN, parent = torch.class('onmt.RVNN', 'nn.Container')

function RVNN:__init (outSize, relDim, numRel, dropout)
  parent.__init(self)
  self.outSize = outSize
  self.relDim = relDim
  self.numRel = numRel
  self.dropout = dropout
  self.net = self:_buildModel()
  self:add(self.net)
end

function RVNN:_buildModel ()
  local model = nn.Linear(self.outSize*2+self.relDim, self.outSize, true)
  local emb = nn.LookupTable(self.numRel, self.relDim)
  local inputs = {nn.Identity()(), nn.Identity()(), nn.Identity()()}
  local rel = emb(inputs[3])
  local proj = nn.JoinTable(2)({inputs[1], inputs[2], rel})
  if self.dropout > 0 then
    proj = onmt.BayesianDropout(self.dropout, 'recursive')(proj)
  end
  local out = nn.Tanh()(model(proj))
  return nn.gModule(inputs, {out})
end

function RVNN:updateOutput(input)
  self.output = self.net:updateOutput(input)
  return self.output
end

function RVNN:updateGradInput(input, gradOutput)
  return self.net:updateGradInput(input, gradOutput)
end

function RVNN:accGradParameters(input, gradOutput, scale)
  return self.net:accGradParameters(input, gradOutput, scale)
end
```

But I found it returns gradient with zero dimension.
I have to change the updateGradInput function to 

```lua
function RVNN:updateGradInput(input, gradOutput)
  self.gradInput = self.net:updateGradInput(input, gradOutput)
  return self.gradInput
end
```

which is not necessary in LSTM.lua.
I can't find any difference between sequencer with LSTM and sequencer with my recursive nets.
I wondering in current sequencer implementation, how self.gradInput is redirected to self.net.gradInput ?
Thanks.

**Comment by guillaumekln** [comment]
How do you initialize the Sequencer?

**Comment by guillaumekln** [comment]
The difference is that you call `backward` on the RVNN module because it is the one exposed by the Sequencer. As you don't override the `backward` function, the definition from `nn.Module` is used:

```lua
function Module:backward(input, gradOutput, scale)
   scale = scale or 1
   self:updateGradInput(input, gradOutput)
   self:accGradParameters(input, gradOutput, scale)
   return self.gradInput
end
```

which expects `self.gradInput` to be not `nil`.

On the other hand, the LSTM module is not directly exposed by the Sequencer and it only relies on `updateGradInput`'s return value. See https://github.com/torch/nngraph/blob/master/gmodule.lua#L420 which is called on each node in the graph.

---

However, these lines:

```lua
self.gradInput = self.net:updateGradInput(input, gradOutput)
return self.gradInput
```

must also appear in the LSTM module for consistency.

So thank you for your question.

**Comment by helson73** [comment]
@guillaumekln That's really helpful. Thanks!
